README.md
# Ceps

Um projeto Flutter!

## Getting Started

Esse projeto tem como objetivo o aprendizado da framework [Flutter](https://github.com/flutter/flutter)

Para ajuda ao iniciar projetos em Flutter, visite a
[documentação online](https://docs.flutter.dev/), que oferece tutoriais, exemplos, guias em desenvolvimento mobile, e uma referência completa na API.

# Instalação
> Para executar este projeto, verifique os requisitos de instalação do Flutter na documentação online.

- Faça o clone  desse projeto em algum diretório no seu computador
- Abra um terminal neste diretório e execute o comando
```
flutter emulators --launch <id_do_emulador>
``` 
- Após inicializar o emulador, execute o projeto com o comando 
```
flutter run
```

# Documentação
Este projeto tem como formato um aplicativo de celular que busca um CEP utilizando o webservice [ViaCep](https://viacep.com.br)

- O usuário informa um  CEP, o aplicativo então faz uma  requisição para o web service e este retorna o resultado do cep nos formatos requeridos pelo usuário `JSON`, `JSONP` ou `XML`

# Exemplos


CHANGELOG.md
# Changelog CEPs

## Version: 1.0
## Features:
- Criação do projeto e configuração inicial

test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:ceps/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/main.dart
import 'package:flutter/material.dart';

//Packages
import 'package:ceps/apps/index.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(
          scaffoldBackgroundColor: const Color(0xffCCCCCC),
          colorScheme: ColorScheme.fromSwatch().copyWith(
            primary: const Color(0xff006600),
            secondary: const Color(0xff919106),
          ),
          inputDecorationTheme: InputDecorationTheme(
              border: OutlineInputBorder(
                  borderRadius: BorderRadius.all(Radius.circular(32))),
              fillColor: MaterialStateColor.resolveWith(
                  (states) => Color.fromARGB(255, 230, 230, 230)),
              filled: true,
              hintStyle: TextStyle(color: Color.fromARGB(255, 163, 163, 163))),
          radioTheme: RadioThemeData(
              fillColor: MaterialStateColor.resolveWith(
                  (states) => const Color(0xff006600))),
          textTheme:
              const TextTheme(bodyMedium: TextStyle(color: Colors.white)),
        ),
        home: const Index());
  }
}


lib/apps/index.dart
import 'package:ceps/apps/search.dart';
import 'package:flutter/material.dart';

class Index extends StatelessWidget {
  const Index({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Container(
      decoration: BoxDecoration(
          image: DecorationImage(
              fit: BoxFit.none,
              alignment: Alignment.bottomLeft,
              image: new AssetImage('assets/Vectorback.png'))),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Container(
                  height: 100,
                  child:
                      Image.asset('assets/Vectorlogo.png', fit: BoxFit.fill)),
              Container(
                margin: const EdgeInsets.only(top: 400),
                height: 65,
                width: 245,
                child: ElevatedButton(
                    onPressed: () => {
                          Navigator.push(
                            context,
                            MaterialPageRoute(
                                builder: (context) => const Search()),
                          )
                        },
                    child: const Text('Pesquisar CEPs')),
              ),
            ],
          )
        ],
      ),
    ));
  }
}


lib/apps/search.dart
import 'package:ceps/apps/result.dart';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

enum DataFormat { json, jsonp, xml }

class Search extends StatefulWidget {
  const Search({super.key});

  @override
  State<Search> createState() => _Search();
}

class _Search extends State<Search> {
  DataFormat? _format = DataFormat.json;

  @override
  Widget build(BuildContext context) {
    late Future<String> futureCep;
    TextEditingController cepInput = TextEditingController();

    double height = MediaQuery.of(context).size.height;
    double width = MediaQuery.of(context).size.width;
    return Scaffold(
        body: SingleChildScrollView(
      child: Stack(children: [
        Container(
            margin: EdgeInsets.only(top: (height * 0.25)),
            child: Center(child: Image.asset('assets/Vectorlogo.png'))),
        Container(
          margin: EdgeInsets.only(top: (height * 0.77)),
          decoration: BoxDecoration(
            color: Theme.of(context).colorScheme.secondary,
            borderRadius: const BorderRadius.only(
                topRight: Radius.circular(32.0),
                topLeft: Radius.circular(32.0)),
          ),
          child: Container(
            margin:
                const EdgeInsetsDirectional.only(top: 40, start: 16, end: 16),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Expanded(
                      child: Container(
                        margin: const EdgeInsets.only(right: 24),
                        height: 56,
                        child: TextField(
                          controller: cepInput,
                          keyboardType: TextInputType.number,
                          decoration: InputDecoration(
                              labelText: 'CEP', hintText: '00.000-000'),
                        ),
                      ),
                    ),
                    SizedBox(
                      height: 56,
                      child: ElevatedButton(
                          onPressed: () => {
                                // futureCep = fetchCEP(),
                                Navigator.push(
                                  context,
                                  MaterialPageRoute(
                                      builder: (context) => const Result()),
                                )
                              },
                          child: const Text('Pesquisar')),
                    ),
                  ],
                ),
                Container(
                  margin: const EdgeInsets.only(top: 30, bottom: 10),
                  child: const Text('Selecione o formato da resposta:',
                      textAlign: TextAlign.left),
                ),
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceAround,
                  children: [
                    Expanded(
                      flex: 1,
                      child: Row(
                        children: [
                          Radio<DataFormat>(
                              value: DataFormat.json,
                              groupValue: _format,
                              onChanged: (DataFormat? value) {
                                setState(() {
                                  _format = value;
                                });
                              }),
                          const Expanded(
                            child: Text('JSON'),
                          )
                        ],
                      ),
                    ),
                    Expanded(
                      flex: 1,
                      child: Row(
                        children: [
                          Radio<DataFormat>(
                              value: DataFormat.jsonp,
                              groupValue: _format,
                              onChanged: (DataFormat? value) {
                                setState(() {
                                  _format = value;
                                });
                              }),
                          const Expanded(
                            child: Text('JSONP'),
                          )
                        ],
                      ),
                    ),
                    Expanded(
                      flex: 1,
                      child: Row(
                        children: [
                          Radio<DataFormat>(
                              value: DataFormat.xml,
                              groupValue: _format,
                              onChanged: (DataFormat? value) {
                                setState(() {
                                  _format = value;
                                });
                              }),
                          const Expanded(
                            child: Text('XML'),
                          )
                        ],
                      ),
                    ),
                  ],
                ),
              ],
            ),
          ),
        ),
      ]),
    ));
  }

  // Future<http.Response> fetchCEP(String cep) async{
  //   String cep =
  //   String url = "https://viacep.com.br/ws/$/json/";

  //   final response = await http.get(Uri.parse(url));

  //   return response;
  // }
}


linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "ceps");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "ceps");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.Create(L"ceps", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


lib/apps/result.dart
import 'package:flutter/material.dart';

enum DataFormat { json, jsonp, xml }

class Result extends StatefulWidget {
  const Result({super.key});

  @override
  State<Result> createState() => _Result();
}

class _Result extends State<Result> {
  DataFormat? _format = DataFormat.json;

  @override
  Widget build(BuildContext context) {
    double height = MediaQuery.of(context).size.height;
    double width = MediaQuery.of(context).size.width;
    return Scaffold(
      appBar: AppBar(
        title: const Text('Resultado da Pesquisa'),
      ),
      body: Column(children: [
        Container(
          decoration: BoxDecoration(
            color: Theme.of(context).colorScheme.secondary,
            borderRadius: const BorderRadius.only(
                bottomLeft: Radius.circular(32.0),
                bottomRight: Radius.circular(32.0)),
          ),
          child: Container(
            height: 180,
            margin:
                const EdgeInsetsDirectional.only(top: 20, start: 16, end: 16),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Expanded(
                      child: Container(
                        margin: const EdgeInsets.only(right: 24),
                        height: 56,
                        child: const TextField(
                          keyboardType: TextInputType.number,
                          decoration: InputDecoration(
                              labelText: 'CEP', hintText: '00.000-000'),
                        ),
                      ),
                    ),
                    SizedBox(
                      height: 56,
                      child: ElevatedButton(
                          onPressed: () => {
                                Navigator.push(
                                  context,
                                  MaterialPageRoute(
                                      builder: (context) => const Result()),
                                )
                              },
                          child: const Text('Pesquisar')),
                    ),
                  ],
                ),
                Container(
                  margin: const EdgeInsets.only(top: 30, bottom: 10),
                  child: const Text('Selecione o formato da resposta:',
                      textAlign: TextAlign.left),
                ),
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceAround,
                  children: [
                    Expanded(
                      flex: 1,
                      child: Row(
                        children: [
                          Radio<DataFormat>(
                              value: DataFormat.json,
                              groupValue: _format,
                              onChanged: (DataFormat? value) {
                                setState(() {
                                  _format = value;
                                });
                              }),
                          const Expanded(
                            child: Text('JSON'),
                          )
                        ],
                      ),
                    ),
                    Expanded(
                      flex: 1,
                      child: Row(
                        children: [
                          Radio<DataFormat>(
                              value: DataFormat.jsonp,
                              groupValue: _format,
                              onChanged: (DataFormat? value) {
                                setState(() {
                                  _format = value;
                                });
                              }),
                          const Expanded(
                            child: Text('JSONP'),
                          )
                        ],
                      ),
                    ),
                    Expanded(
                      flex: 1,
                      child: Row(
                        children: [
                          Radio<DataFormat>(
                              value: DataFormat.xml,
                              groupValue: _format,
                              onChanged: (DataFormat? value) {
                                setState(() {
                                  _format = value;
                                });
                              }),
                          const Expanded(
                            child: Text('XML'),
                          )
                        ],
                      ),
                    ),
                  ],
                ),
              ],
            ),
          ),
        ),
        Container(
          width: width,
          decoration: const BoxDecoration(
            color: Color(0xff539103),
            borderRadius: BorderRadius.only(
                topLeft: Radius.circular(32.0),
                topRight: Radius.circular(32.0),
                bottomLeft: Radius.circular(32.0),
                bottomRight: Radius.circular(32.0)),
          ),
          padding: const EdgeInsets.all(24),
          margin: const EdgeInsets.all(16),
          child: Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  'CEP:',
                ),
                SizedBox(height: 15),
                Text('Formato:'),
                SizedBox(height: 15),
                Text(
                    '{ "cep": "76801-799",  "logradouro": "Avenida dos Imigrantes",  "complemento": "até 465 - lado ímpar",  "bairro": "São Sebastião",  "localidade": "Porto Velho",  "uf": "RO",  "ibge": "1100205",  "gia": "",  "ddd": "69",  "siafi": "0003"}')
              ],
            ),
          ),
        ),
      ]),
    );
  }
}
